contracts/RewardDelegate.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.7.6;
+
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { IRewardDelegate } from "./interface/IRewardDelegate.sol";
+
+contract RewardDelegate is IRewardDelegate {
+    using Address for address;
+    using SafeMath for uint256;
+
+    // truster => beneficiaryCandidate
+    mapping(address => address) internal _beneficiaryCandidateMap;
+    // truster => beneficiary
+    mapping(address => address) internal _beneficiaryMap;
+    // beneficiary => how many trusters explicitly delegate to this beneficiary,
+    // excluding beneficiary self
+    // explicitly means calling setBeneficiaryCandidate() and updateBeneficiary()
+    mapping(address => uint256) internal _trusterCountMap;
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    function setBeneficiaryCandidate(address candidate) external override {
+        address truster = msg.sender;
+
+        // RD_TNC: truster is not a contract
+        require(truster.isContract(), "RD_TNC");
+
+        // RD_CE: candidate error
+        require(candidate != truster && !candidate.isContract() && candidate != address(0), "RD_CE");
+
+        _beneficiaryCandidateMap[truster] = candidate;
+    }
+
+    function updateBeneficiary(address truster) external override {
+        address beneficiary = msg.sender;
+
+        // RD_CNS: candidate not set
+        require(_beneficiaryCandidateMap[truster] == beneficiary, "RD_CNS");
```

Comment:
should clear `_beneficiaryCandidateMap[truster]` after check to avoid double-update (it might mess up the counts as well)

Reply:
per discussion, also need to require `_beneficiaryMap[truster] != beneficiary`